packages/hub/src/lib/parse-safetensors.ts

```diff
@@ -0,0 +1,161 @@
+	const lengthOfHeader = new DataView(bufLengthOfHeaderLE).getBigUint64(0, true);
+	/// ^little-endian
+
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const secondResp = (await downloadFile({ ...params, path, range: [8, 7 + Number(lengthOfHeader)] }))!;
+	const header: FileHeader = await secondResp.json();
+	/// no validation for now, we assume it's a valid FileHeader.
+	return header;
+}
+
+async function parseShardedIndex(
+	path: string,
+	params: {
+		repo: RepoDesignation;
+		revision?: string;
+		credentials?: Credentials;
+		hubUrl?: string;
+	}
+): Promise<{ index: IndexJson; headers: ShardedHeaders }> {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const indexResp = (await downloadFile({
+		...params,
+		path,
+	}))!;
+	const index: IndexJson = await indexResp.json();
+	/// no validation for now, we assume it's a valid IndexJson.
+
+	const shardedMap: ShardedHeaders = {};
+	const filenames = [...new Set(Object.values(index.weight_map))];
+	await Promise.all(
+		filenames.map(async (filename) => {
+			shardedMap[filename] = await parseSingleFile(filename, params);
+			/// Note(insertion order is not deterministic)
+		})
+	);
+	return { index, headers: shardedMap };
+}
+
+export async function parseSafetensorsFromModelRepo(params: {
+	repo: RepoDesignation;
+	hubUrl?: string;
+	credentials?: Credentials;
+	revision?: string;
+}): Promise<ParseFromRepo> {
+	checkCredentials(params.credentials);
+	const repoId = toRepoId(params.repo);
+
+	if (repoId.type !== "model") {
+		throw new TypeError("Only model repos should contain safetensors files.");
+	}
+
+	if (await fileExists({ ...params, path: SINGLE_FILE })) {
+		return {
+			sharded: false,
+			header: await parseSingleFile(SINGLE_FILE, params),
+		};
+	} else if (await fileExists({ ...params, path: INDEX_FILE })) {
+		return {
+			sharded: true,
+			...(await parseShardedIndex(INDEX_FILE, params)),
+		};
+	} else {
+		throw new Error("model id does not seem to contain safetensors weights");
+	}
+}
+
+function computeNumOfParamsByDtypeSingleFile(header: FileHeader): Counter<Dtype> {
+	const n = new Counter<Dtype>();
+	const tensors = omit(header, "__metadata__");
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	for (const [k, v] of typedEntries(tensors)) {
+		if (v.shape.length === 0) {
+			continue;
+		}
+		n.incr(
+			v.dtype,
+			v.shape.reduce((a, b) => a * b)
+		);
+	}
+	return n;
+}
+
+function computeNumOfParamsByDtypeSharded(shardedMap: ShardedHeaders): Counter<Dtype> {
+	const n = new Counter<Dtype>();
+	for (const v of Object.values(shardedMap)) {
+		n.add(computeNumOfParamsByDtypeSingleFile(v));
+	}
+	return n;
+}
+
+export function computeNumOfParamsByDtype(parse: ParseFromRepo): Counter<Dtype> {
+	if (parse.sharded) {
+		return computeNumOfParamsByDtypeSharded(parse.headers);
+	} else {
+		return computeNumOfParamsByDtypeSingleFile(parse.header);
+	}
+}
```

Comment:
Imo it's too generic a function name to be exported (when it's not part of a safetensors namespace/module)

Maybe if we keep the function we can just compute it always and include it in the result of `ParseFromRepo`?

Also, exporting a `Counter` class to the users of the lib is not great, we should export the results of `Counter.dict()` or even better, export a `Record<DType, number>` eg:

```
{
  BOOL: 0,
  I8: 0,
  I16: 21892,
  I32: 0,
  ...
}
```

Not sure we need to add a `total` field, a user could just `sum(Object.values(X))`

Reply:
> Maybe if we keep the function we can just compute it always and include it in the result of `ParseFromRepo`?

I'd rather not, as i prefer the output of parse be the exact data from the files

Reply:
Ok i will remove `Counter` completely

Reply:
3757fc3598049bf3eba451e5e363a958f00f25b3